### PR TITLE
errors: Introduce error code handling

### PR DESCRIPTION
**Pull request by kbeckmann**
This is a draft to show what I mean in issue #11

### CODE REVIEW

**Comment by kbeckmann** [comment]
In GDB i can do this now:

```
0x080062dc in main () at app/main.c:120
120		ERR_CHECK(r);
(gdb) print g_status
$1 = HAL_ERROR
```

File: app/drivers/adc.h [hunk 1]
```diff
@@ -1,5 +1,9 @@
 #pragma once
 
 #include "stm32l4xx_hal.h"
+#include "errors.h"
 
-HAL_StatusTypeDef adc_init(void);
+#define ADC_INIT_FAIL	(ADC_BASE + 0)
+#define ADC_CONFIG_FAIL	(ADC_BASE + 1)
```
Comment (arturo182) [inline app/drivers/adc.h]: Not sure about this FAIL suffix, maybe just do a E prefix like EADC_INIT

File: app/errors.c [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#include "errors.h"
+
+HAL_StatusTypeDef g_status;
```
Comment (arturo182) [inline app/errors.c]: Hmm, hal_error, hal_status?

File: app/drivers/adc.h (hunk 1 above)
Comment (kbeckmann) [reply app/drivers/adc.h]: Agreed. EADC_INIT etc is better.

File: app/errors.c (hunk 2 above)
Comment (kbeckmann) [reply app/errors.c]: g_hal_error ?

Comment (kbeckmann) [reply app/errors.c]: well it's a global after all 🌍

File: app/drivers/adc.h (hunk 1 above)
Comment (kbeckmann) [reply app/drivers/adc.h]: It's when HAL_ADC_Init fails, so can't really come up with a better name. EADC_HAL_INIT?

**Comment by kbeckmann** [comment]
Got a crazy idea when doing this. What about reserving the least 2 significant bits for HAL_StatusTypeDef? 
```
typedef enum
{
  HAL_OK       = 0x00,
  HAL_ERROR    = 0x01,
  HAL_BUSY     = 0x02,
  HAL_TIMEOUT  = 0x03
} HAL_StatusTypeDef;
```

That way we can have good errors, but still encode the HAL error without the need of a global, and no worries about race conditions in case you want to check the hal error code programatically.

**Comment by kbeckmann** [comment]
Updated the branch with above changes. Think it works out pretty well.

File: app/drivers/i2c.c [hunk 3]
```diff
@@ -57,12 +54,22 @@ void HAL_I2C_MspDeInit(I2C_HandleTypeDef *p_handle)
...
-HAL_StatusTypeDef i2c_master_tx(uint16_t dev_address, uint8_t *p_data, uint16_t size, uint32_t timeout_ms)
+err_t i2c_master_tx(uint16_t dev_address, uint8_t *p_data, uint16_t size, uint32_t timeout_ms)
 {
-	return HAL_I2C_Master_Transmit(&i2c_handle, dev_address, p_data, size, timeout_ms);
+	HAL_StatusTypeDef status;
+
+	status = HAL_I2C_Master_Transmit(&i2c_handle, dev_address, p_data, size, timeout_ms);
+	HAL_ERR_CHECK(status, EI2C_HAL_MASTER_TRANSMIT);
+
+	return status;
```
Comment (arturo182) [inline app/drivers/i2c.c]: Should return ERROR_OK

File: app/drivers/i2c.c [hunk 4]
```diff
@@ -18,16 +18,13 @@ HAL_StatusTypeDef i2c_init(void)
...
-	if (status != HAL_OK)
-		return status;
+	HAL_ERR_CHECK(status, EI2C_HAL_CONFIG_ANALOG_FILTER);
 
 	status = HAL_I2CEx_ConfigDigitalFilter(&i2c_handle, 0);
-	if (status != HAL_OK)
-		return status;
+	HAL_ERR_CHECK(status, EI2C_HAL_CONFIG_DIGITAL_FILTER);
 
 	return status;
```
Comment (arturo182) [inline app/drivers/i2c.c]: Here too

File: app/drivers/i2c.c [hunk 5]
```diff
@@ -57,12 +54,22 @@ void HAL_I2C_MspDeInit(I2C_HandleTypeDef *p_handle)
...
-HAL_StatusTypeDef i2c_master_rx(uint16_t dev_address, uint8_t *p_data, uint16_t size, uint32_t timeout_ms)
+err_t i2c_master_rx(uint16_t dev_address, uint8_t *p_data, uint16_t size, uint32_t timeout_ms)
 {
-	return HAL_I2C_Master_Receive(&i2c_handle, dev_address, p_data, size, timeout_ms);
+	HAL_StatusTypeDef status;
+
+	status = HAL_I2C_Master_Receive(&i2c_handle, dev_address, p_data, size, timeout_ms);
+	HAL_ERR_CHECK(status, EI2C_HAL_MASTER_RECEIVE);
+
+	return status;
```
Comment (arturo182) [inline app/drivers/i2c.c]: Here

**Pull request merged**